### DISCUSSION TITLE
关于灰度测试「弹幕智能云屏蔽」

### DISCUSSION THREAD

**Discussion by WakelessSloth56** [the1812/Bilibili-Evolved#4920] [综合讨论 / General]
> 参见讨论 https://github.com/the1812/Bilibili-Evolved/discussions/4828 。

目前「弹幕智能云屏蔽」同时存在一下三个版本：

<details><summary>1. 灰度测试前的版本</summary>

![20240919232854](https://github.com/user-attachments/assets/1c929b29-eb6d-46e2-bf7b-ac9723d7c13b)
就是很单纯的判断「智能云屏蔽」的开关和等级，没什么好说的。
```javascript
t.prototype.aiJudge = function (e, t) {
    return !!(t.aiSwitch && Math.abs(e.weight) < t.aiLevel);
};
```

</details>

<details><summary>2. 灰度测试早期的版本</summary>

![20240919233000](https://github.com/user-attachments/assets/fcacb780-224c-43b1-87a2-da64952d5929)
加入了对 `dmAiBlockAbTestFlag` 的判断，如果为 `false` 的话就是原先的逻辑，否则的话屏蔽等级最低都是 `3`。
而 `dmAiBlockAbTestFlag` 来自 `window.webAbTest.danmuku_block_version === 'NEW'`。
```javascript
Object.defineProperty(t.prototype, "dmAiBlockAbTestFlag", {
    get: function() {
        var e;
        return "NEW" === (null === (e = window.webAbTest) || void 0 === e ? void 0 : e.danmuku_block_version)
    },
    enumerable: !1,
    configurable: !0
})
// ...
t.prototype.aiJudge = function (e, t) {
    if (((this.totalFiltleredDm += 1), this.dmAiBlockAbTestFlag)) {
        if (Math.abs(e.weight) < Math.max(t.aiLevel, 3))
            return (this.lowWeightDm += 1), !0;
    } else if (t.aiSwitch && Math.abs(e.weight) < t.aiLevel) return !0;
    return !1;
};
```

</details>

<details><summary>3. 近期又出现的新版本</summary>

![20240919233049](https://github.com/user-attachments/assets/3cc825bb-4afe-43e6-bfd9-a097bd0ad336)
`aiJudge` 里没有了对 `dmAiBlockAbTestFlag` 的判断，但是传入的“等级”参数 `aiLevel`，却是来自于弹幕显示区域和弹幕密度。
```javascript
Object.defineProperty(t.prototype, 'aiLevel', {
    get: function () {
        return this.state.dmarea < 100 && this.state.dmarea > 0
            ? 3
            : this.state.dmdensity === n.most || this.state.dmdensity === n.more
            ? 2
            : this.state.dmdensity === n.normal
            ? 3
            : void 0;
    },
    enumerable: !1,
    configurable: !0,
});
// ...
t.prototype.isBlockDanmaku = function (e, t) {
    // ...
    if (!t) {
        if (this.aiJudge(e, this.dmSettingStore.aiLevel)) return !0;
        // ...
    }
    // ...
};
// ...
t.prototype.aiJudge = function(e, t) {
  return this.totalFiltleredDm += 1,
  Math.abs(e.weight) < t
}
```
</details>

---

综合上面的分析：

```javascript
// ==UserScript==
// @name         禁用弹幕智能云屏蔽
// @namespace    https://github.com/the1812/Bilibili-Evolved/discussions/4920
// @version      2024-10-21
// @description  禁用弹幕智能云屏蔽
// @author       WakelessSloth56
// @match        https://www.bilibili.com/video/*
// @match        https://www.bilibili.com/bangumi/play/*
// @grant        none
// @run-at       document-start
// ==/UserScript==
(function () {
    "use strict";
    const interval = setInterval(() => {
        if (window.player) {
            window.player.on(
                window.nano.EventType.Player_Play,
                () => {
                    const dmSetting = JSON.parse(localStorage.bpx_player_profile).dmSetting;
                    dmSetting.aiSwitch = false; // 针对版本 2
                    dmSetting.dmarea = 100; // 针对版本 3
                    dmSetting.dmdensity = 0; // 针对版本 3
                    window.player.setDanmakuSetting(dmSetting);
                    window.webAbTest.danmuku_block_version = "OLD"; // 针对版本 2
                }
            );
            clearInterval(interval);
        }
    }, 100);
})();
```

- 如果发现脚本不生效，请看一下脚本版本（`@version`）的日期，如果你在用的旧的话可以更新一下再试试。
- 对于版本 2，使用此脚本虽然在控件上还是没法看到开关，但是实际上已经关了。
- 对于版本 3，此脚本会导致控件中的「显示区域」和「弹幕密度」不再起作用。
- 由于目前我并没有摸清楚会使用哪个版本取决于什么，我测试的时候已经是版本 3 了，但是偶尔还是会出现版本 2，所以也存在此脚本也无效的情况……那只能继续再讨论了……

---

其他方法：
 - [Pakku v2024.10.1](https://github.com/xmcp/pakku.js/releases/tag/2024.10.1) 已更新【阻止弹幕云屏蔽】功能（ https://github.com/the1812/Bilibili-Evolved/discussions/4920#discussioncomment-10778418  @xmcp）。**如果已经在使用 Pakku 或不介意安装浏览器扩展的话这个方法更好。**

**Comment by lainio24** [comment]
https://www.bilibili.com/read/cv36036480/

好好好，不会是 B 站官方看到这个玩意才又改出了版本3吧（）

**Reply by lainio24** [reply]
还有你 B 的神人程序员，`window.webAbTest.danmuku_block_version`，`danmaku`？`danmuku`😅

**Reply by Anderson2001-js** [reply]
关于danmuku这个，详情可见artplayer

**Comment by SSpinachZ** [comment]
我现在是版本3，用这个脚本不生效，但用bilibili旧播放页那种替换整个播放器的脚本可以看到所有弹幕。
可能b站还有其他判定条件？

**Reply by WakelessSloth56** [reply]
替换整个播放器肯定能看到所有弹幕啊……
版本 3 的话，用这个脚本的还得注意一下「显示区域」要设成 `100%`。(我已经更新脚本了)

**Reply by GalaxySnail** [reply]
显示区域设置成 100% 的话弹幕遮挡画面就太多了，我以往都是 25%，请问是否有可能不修改这个属性而是通过其它方式绕过？

**Reply by WakelessSloth56** [reply]
没有，除非是直接修改播放器（比如[Bilibili旧播放页](https://greasyfork.org/zh-CN/scripts/394296)）。弹幕显示区域不是 100% 的话就是 3 级屏蔽。

**Reply by GalaxySnail** [reply]
有可能通过 monkey patch 相关逻辑的函数来解决吗？

**Comment by doubilbm** [comment]
我现在是版本3，有一部分弹幕可以显示出来了，但对比弹幕列表还是不全，是不是还有别的屏蔽选项？

**Reply by WakelessSloth56** [reply]
这个只针对「智能云屏蔽」，不影响别的屏蔽。

**Reply by doubilbm** [reply]
> 这个只针对「智能云屏蔽」，不影响别的屏蔽。

用了这个脚本弹幕是比以前多了一点点，但是还有一部分弹幕依旧显示不了，所以我是想问一下除了 智能云屏蔽 是不是还有其它的屏蔽机制，或者是这个脚本并没有完全禁用云屏蔽，只是恢复成原来的1级屏蔽？
我看视频从来不开任何弹幕屏蔽，这段时间真是给我恶心坏了，之前那个版本2有个脚本能改回去，现在只能被迫用客户端看视频......

**Reply by WakelessSloth56** [reply]
「智能云屏蔽」是一定能关掉的。具体能给个示例吗，哪个视频的哪条弹幕没有显示？

**Comment by xmcp** [comment]
好家伙，原来版本3并不是删掉了云屏蔽，而是藏到其他选项里面了。

建议直接用 [pakku的用户脚本](https://github.com/xmcp/pakku.js/tree/master/userscript_docs) 把所有弹幕的权重都改成11，这样不管叔叔再怎么改都不会受影响了。

```javascript
tweak_after_pakku(c=>{
  for(let d of c.objs)
    d.weight = 11;
});
```

**Comment by xmcp** [comment]
[pakku v2024.10.1](https://github.com/xmcp/pakku.js/releases/tag/2024.10.1) 已经增加了【阻止弹幕云屏蔽】功能，更新后可以直接在选项页面启用，不再需要上面贴的用户脚本。因为它的原理是覆盖弹幕权重，应该对目前所有播放器版本都有效且无副作用。

另外此版本还改进了pakku的【自动弹幕优选】逻辑，此功能将按照实时弹幕密度来调整云屏蔽等级，可以视为播放器的智能云屏蔽功能的上位替代。

**Reply by WakelessSloth56** [reply]
还得是浏览器扩展啊，我当时有想过修改弹幕的权重，但是现在播放器又不暴露弹幕列表了，用户脚本劫持弹幕 API 的响应也不太好搞，就去研究了播放器的代码整出来了上面的脚本。
早期还有这种 https://github.com/the1812/Bilibili-Evolved/discussions/4828#discussioncomment-10640462 歪门邪道的方法（笑

**Comment by GalaxySnail** [comment]
在B站刷到了一篇专栏 [cv39639593](https://www.bilibili.com/read/cv39639593/)，是写了一个浏览器拓展拦截 HTTP 请求修改 js 从而返回旧版。不过油猴脚本应该是没法这么做了。

（现在未登录的情况下直接访问传统的专栏 URL 也会被重定向到新版 opus 页面，真恶心啊）

**Reply by WakelessSloth56** [reply]
拦截请求油猴也可以的，我之前就搞过这种方案 https://github.com/the1812/Bilibili-Evolved/discussions/4828#discussioncomment-10640462

**Reply by GalaxySnail** [reply]
这个实现只能拦 `<script>` 标签吧，没法拦 require 和 import

**Reply by WakelessSloth56** [reply]
是。但是B站播放器也就是一个 `script`：
```html
<script type='text/javascript' src='//s1.hdslb.com/bfs/static/player/main/core.3117538b.js'></script>
```
